### PR TITLE
feat: add Obico for Klipper to KIAUH

### PR DESCRIPTION
**Pull request by kennethjiang**
Hi @th33xitus, I think this PR is ready for review now. Sorry for the delay.

Here are a few areas that I will need your feedback on:

1. I have been naming the systemd service after the pattern of `moonraker-obico[-xxx].service`, which caused some problem with existing code. Since moonraker-obico has gone live for a while, I don't want to break the compatibility with the users already using it, I had to make changes [here](https://github.com/th33xitus/kiauh/pull/227/files#diff-51c13f857871ab2262217665d19263f641f7765a3047b917380c56e6271a0c0bL682) and [here](https://github.com/th33xitus/kiauh/pull/227/files#diff-ba13e70564a03881545708297775e45ad530fe9daf77c3a4e7577f88c2d75fc8L20). The 2nd one is quite ugly. Let me know if you have a better idea.
2. `moonraker-obico` can be in an *installed* but *not linked" status. I made the change [here](https://github.com/th33xitus/kiauh/pull/227/files) and [here](https://github.com/th33xitus/kiauh/pull/227/files#diff-4a291dc3dd20a7014d941b94a8610be1e17eda608a475fe91514b80cc79a8475R15). Again this is a bit of inverted dependency because now `main_menu.sh` and `install_menu.sh` depend on functions in `obico.sh`. Will love to hear your feedback here too.

Screenshots for single-instance installation:

Screen Shot 2022-08-01 at 3 17 22 PM
Screen Shot 2022-08-01 at 3 17 46 PM
Screen Shot 2022-08-01 at 3 17 32 PM
Screen Shot 2022-08-01 at 3 18 03 PM

Screenshots for multi-instance installation:

Screen Shot 2022-08-01 at 6 21 59 PM
Screen Shot 2022-08-01 at 6 24 55 PM

### CODE REVIEW

**Comment by th33xitus** [comment]
Hi @kennethjiang, on first glance it looks like its going into the right direction. I saw one or two oddities, but as you stated that this is a rough draft, i'll wait until you give the go for me to start the formal review.

For example: 
* i don't think that we need a python 3 check, as obico requires moonraker and moonraker requires python 3. 
* also im not a fan of using the `.*` wildcard in the regex here: https://github.com/th33xitus/kiauh/pull/227/files#diff-08e65de842c3dbbcc98be9e5467e5447241080fbb0bcb661f162d58df16e6a60R20
* this is probably just kind of a copy & paste "placeholder"? https://github.com/th33xitus/kiauh/pull/227/files#diff-08e65de842c3dbbcc98be9e5467e5447241080fbb0bcb661f162d58df16e6a60R242

File: scripts/obico.sh [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+    moonraker_obico_names+=( "$(get_instance_name "${service}" moonraker-obico)" )
+  done
+
+  local allowed_moonraker_obico_count=$(( moonraker_count - existing_moonraker_obico_count ))
+  if (( allowed_moonraker_obico_count > 0 )); then
+
+    ### Step 1: Ask for the number of moonraker-obico instances to install
+    if (( moonraker_count == 1 )); then
+      ok_msg "Moonraker installation found!\n"
+      new_moonraker_obico_count=1
```
Comment (th33xitus) [inline scripts/obico.sh:101]:
```ssh 
new_moonraker_obico_count=1
```
I'm unsure about the usage of this variable. It seems to be not used anywhere? Is it used somewhere else as it is not declared as a local variable?

File: scripts/obico.sh [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+
+    ### Step 4: Install dependencies
+    local dep=(git dfu-util virtualenv python3 python3-pip python3-venv ffmpeg)
+    dependency_check "${dep[@]}"
+
+    ### Step 5: Clone the moonraker-obico repo
+    clone_or_update_moonraker_obico "${MOONRAKER_OBICO_REPO}"
+
+    ### step 6: call moonrake-obico/install.sh with the correct params
+    local log="${KLIPPER_LOGS}"
```
Comment (th33xitus) [inline scripts/obico.sh:186]:
```ssh
local log="${KLIPPER_LOGS}"
```
This variable seems to be unused?

File: scripts/obico.sh (hunk 1 above)
Comment (kennethjiang) [reply scripts/obico.sh:101]: Good catch. It's a bug! Do you want me to push this change?

```
diff --git a/scripts/obico.sh b/scripts/obico.sh
index 791fcdb..de2bd3f 100644
--- a/scripts/obico.sh
+++ b/scripts/obico.sh
@@ -98,7 +98,6 @@ function moonraker_obico_setup_dialog() {
     ### Step 1: Ask for the number of moonraker-obico instances to install
     if (( moonraker_count == 1 )); then
       ok_msg "Moonraker installation found!\n"
-      new_moonraker_obico_count=1
     elif (( moonraker_count > 1 )); then
       top_border
       printf "|${green}%-55s${white}|\n" " ${moonraker_count} Moonraker instances found!"
@@ -134,7 +133,7 @@ function moonraker_obico_setup_dialog() {
         (( new_moonraker_obico_count > allowed_moonraker_obico_count )) && error_msg "Number of Moonraker-obico instances larger than installed Moonraker instances"
       done && select_msg "${new_moonraker_obico_count}"
     else
-      log_error "Internal error. new_moonraker_obico_count of '${new_moonraker_obico_count}' not equal or grather than one!"
+      log_error "Internal error. moonraker_count of '${moonraker_count}' not equal or grather than one!"
       return 1
     fi
```

File: scripts/obico.sh (hunk 2 above)
Comment (kennethjiang) [reply scripts/obico.sh:186]: Another good catch. I guess I planed to use it but ended up using `"${KLIPPER_LOGS}"` directly. Feel free to remove it.

**Comment by kennethjiang** [comment]
@th33xitus I changed this PR based on our discussion and tested all the cases I can think of.

I found 1 bug: on the install_menu screen, if the user selects 1) [klipper], 2) [moonraker], and 9) [Obico for Klipper] without going back to the main screen, function `get_multi_instance_names` will return an empty string.

I did some debugging and found that `.kiauh.ini` is not updated until the user goes back to the main screen. However, I struggled to figure out where I did wrong. Will appreciate it if you can provide some guidance here.

Please also go over the code again, particularly commits 7837a78c4b6113e76dc883058118c0a21eef3410, 7837a78c4b6113e76dc883058118c0a21eef3410, and 8d284179d8a21f7ef6720d5c9388a27e7fce58c5

**Comment by th33xitus** [comment]
> I found 1 bug: on the install_menu screen, if the user selects 1) [klipper], 2) [moonraker], and 9) [Obico for Klipper] without going back to the main screen, function get_multi_instance_names will return an empty string.

I did some debugging and found that .kiauh.ini is not updated until the user goes back to the main screen. However, I struggled to figure out where I did wrong. Will appreciate it if you can provide some guidance here.

You didn't do any wrong. You actually pointed me to a maybe not so smart solution.
https://github.com/th33xitus/kiauh/blob/master/scripts/ui/main_menu.sh#L85

The `init_ini` function is called whenever the main menu is loaded.
https://github.com/th33xitus/kiauh/blob/master/scripts/utilities.sh#L125

And in there the multi instance names are then written to the ini
https://github.com/th33xitus/kiauh/blob/master/scripts/utilities.sh#L189

Maybe it is smarter to factor `fetch_webui_ports` and `fetch_multi_instance_names` out of the `init_ini` function and place them both into the `install_menu` function right after `install_ui`. Both those fetch functions aquire data that is usually only used in installation functions, at least i can't think of any other usecases within the script right now.

**Comment by th33xitus** [comment]
@kennethjiang so i just pushed a commit addressing the bug you mentioned.

I have a view comments to make on other pieces of the code, i quite don't understand yet.

File: scripts/obico.sh [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+  local moonraker_cfg_dirs=($(get_config_folders))
+  if [[ -n "${moonraker_cfg_dirs}" ]]; then
+    echo "${moonraker_cfg_dirs[${i}]}/moonraker-obico.cfg"
+  else
+    echo ""
+  fi
+}
+
+function moonraker_obico_needs_linking() {
+  moonraker_obico_cfg=${1}
```
Comment (th33xitus) [inline scripts/obico.sh:34]: Can this variable be local? Or is it required to be global? I see it only being used inside the `moonraker_obico_needs_linking` function

File: scripts/obico.sh [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+
+      for (( i=1; i <= new_moonraker_obico_count; i++ )); do
+        "${MOONRAKER_OBICO_DIR}/install.sh" -n "${moonraker_names[${j}]}" -C "${moonraker_cfg_dirs[${j}]}/moonraker.conf" -p $((port+j)) -H 127.0.0.1 -l "${KLIPPER_LOGS}" -s -L -S "${obico_server_url}"
+        j=$(( j + 1 ))
+      done && unset j
+    fi # (( moonraker_count == 1 ))
+  fi  # (( new_moonraker_obico_count > 0 ))
+
+  ### Step 7: Link to the Obico server if necessary
+  local instance_name
```
Comment (th33xitus) [inline scripts/obico.sh:187]: This variable seems to be unused now?

File: scripts/obico.sh [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+  local not_linked_instances=()
+  if (( moonraker_count == 1 )); then
+    if moonraker_obico_needs_linking "$(moonraker_obico_config 0)"; then
```
Comment (th33xitus) [inline scripts/obico.sh:190]: The `moonraker_obico_config` function defined in line 24 does not take in any arguments. So the `0` you pass there (and a few lines below the `${i}`) won't get used inside the function. Im not quite sure what the intention was, if the arguments are obsolete or if the `moonraker_obico_config` function is incomplete?

File: scripts/obico.sh [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+      not_linked_instances+=("0")
+    fi
+  elif (( moonraker_count > 1 )); then
+    for (( i=0; i <= moonraker_count; i++ )); do
+      if moonraker_obico_needs_linking "$(moonraker_obico_config ${i})"; then
```
Comment (th33xitus) [inline scripts/obico.sh:195]: See comment above :)

File: scripts/obico.sh [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+          abort_msg "Exiting Moonraker-obico setup ...\n"
+          return;;
+        *)
+          error_msg "Invalid Input!";;
+      esac
+    done
+
+    if (( moonraker_count == 1 )); then
+      status_msg "Link moonraker-obico to the Obico Server..."
+      "${MOONRAKER_OBICO_DIR}/scripts/link.sh" -q -c "$(moonraker_obico_config 0)"
```
Comment (th33xitus) [inline scripts/obico.sh:241]: See comment above :)

File: scripts/obico.sh [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+    elif (( moonraker_count > 1 )); then
+      for i in "${not_linked_instances[@]}"; do
+        local name="${moonraker_names[i]}"
+        status_msg "Link moonraker-obico-${name} to the Obico Server..."
+        "${MOONRAKER_OBICO_DIR}/scripts/link.sh" -q -n "${name}" -c "$(moonraker_obico_config ${i})"
```
Comment (th33xitus) [inline scripts/obico.sh:246]: See comment above :)

File: scripts/obico.sh (hunk 5 above)
Comment (kennethjiang) [reply scripts/obico.sh:190]: This is a really good catch! It it supposed to use `${1}`, not `${i}`. Not sure how my testing passed... Fixing it now.

Comment (kennethjiang) [reply scripts/obico.sh:189]: Should be fixed in [276c123](https://github.com/th33xitus/kiauh/pull/227/commits/276c123d3d6d0d3609e0f4b179b545d7d53f40cd). `moonraker_obico_config` should take an argument, which is the index of a specific instance in `multi_instance_names`

**Pull request merged**